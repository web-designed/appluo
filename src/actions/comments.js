
import moment from 'moment'
import database from '../firebase/firebase'

//*******************************************************
// COMMENTS ACTION GENERATORS
//*******************************************************

   export const addComment = (eventId, comment) => ({
      type: 'ADD_COMMENT',
      eventId,
      comment
   })

   export const startAddComment = (eventId, {commentBody, commenter}) => {
      return(dispatch) => {
         const comment = {
            createdAt: moment().valueOf(),
            commenter,
            commentBody
         }

         return database.ref(`events/${eventId}/comments`).push(comment).then( ref => {
            dispatch(addComment(eventId, {
               ...comment,
               id: ref.key
            }))
         })
      }
   }

   export const removeComment = (eventId, commentId) => ({
      type:'REMOVE_COMMENT',
      eventId,
      commentId
   })

   export const editComment = (eventId, commentId, update) => ({
      type:'EDIT_COMMENT',
      eventId,
      commentId,
      update
   })