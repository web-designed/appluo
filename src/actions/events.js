import uuid from 'uuid'
import moment from 'moment'
import database from '../firebase/firebase'


//*******************************************************
// EVENTS ACTION GENERATORS
//*******************************************************

export const addEvent = (event) => ({
   type:'ADD_EVENT',
   event
})

export const startAddEvent = (eventData = {}) => {
   return (dispatch) => {
      const {
         note = '', 
         cleaner = 'unknown', 
         cleanedAt = 0,
         createdAt = 0,
         place = ''
      } = eventData

      const event = { note, cleaner, cleanedAt, createdAt, place }

      return database.ref('events').push(event).then((ref) => { // add return to chain promises in the tests
         dispatch(addEvent({
            id: ref.key,
            ...event
         }))
      })
   }
}


export const removeEvent = (id) => ({
   type: 'REMOVE_EVENT',
   id
})

export const editEvent = (id, update) => ({
   type: 'EDIT_EVENT', 
   id,
   update
})

export const addComment = (eventId, { commentBody, commenter }) => ({
   type: 'ADD_COMMENT',
   eventId,
   comment: {
      id: uuid(),
      createdAt: moment().valueOf(),
      commenter,
      commentBody
   }
})

export const removeComment = (eventId, commentId) => ({
   type:'REMOVE_COMMENT',
   eventId,
   commentId
})

export const editComment = (eventId, commentId, update) => ({
   type:'EDIT_COMMENT',
   eventId,
   commentId,
   update
})