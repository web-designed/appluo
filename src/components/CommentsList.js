import React from 'react'
import { connect } from 'react-redux'
import CommentsListItem from './CommentListItem'

const CommentsList = ({ events, currentEvent }) => {
   return (
      <div>
         <h1>Comments:</h1>
         {currentEvent.comments.map((comment) => (
            <CommentsListItem 
               commentsId={comment.id}
               createdAt={comment.createdAt}
               commenter={comment.commenter}
               commentBody={comment.commentBody}
            />
         ))}
      </div>
   )
}
const mapStateToProps = ({ events }) => ({
   events: events
})
export default connect(mapStateToProps)(CommentsList)