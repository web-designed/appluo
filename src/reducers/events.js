//*******************************************************
// EVENTS REDUCERS
//*******************************************************

   const eventsReducerDefaultState = []
   const eventsReducer = (state = eventsReducerDefaultState, action) => {
      switch(action.type){
         case 'ADD_EVENT':
            return state.concat(action.event)
         case 'REMOVE_EVENT':
            return state.filter(({ id }) => id !== action.id)
         case 'EDIT_EVENT':
            return state.map((event) => {
               if(event.id === action.id) {
                  return {
                     ...event,
                     ...action.update
                  }
               } else {
                  return event
               }
            })
         case 'ADD_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: [...event.comments, action.comment]
                  }
               } else {
                  return event
               }
            })
         case 'REMOVE_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.filter((comment) => {
                        return comment.id !== action.commentId
                     })
                  }
               } else {
                  return event
               }
            })
         case 'EDIT_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.map((comment) => {
                        if(comment.id === action.commentId){
                           return {
                              ...comment,
                              ...action.update
                           }
                        } else {
                           return comment
                        }
                     }) 
                  }
               } else {
                  return event
               }
            })
         default:
            return state
      }
   }

export default eventsReducer