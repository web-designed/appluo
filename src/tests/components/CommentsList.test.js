import React from 'react'
import { shallow } from 'enzyme'
import CommentsList from '../../components/CommentsList'
import comments from '../fixtures/comments'
import events from '../fixtures/events'

test('should render an empty list correctly', () => {
   const wrapper = shallow(<CommentsList currentEvent={events[0]} />)
   expect(wrapper).toMatchSnapshot()
})