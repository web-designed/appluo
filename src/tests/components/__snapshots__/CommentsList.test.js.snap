// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an empty list correctly 1`] = `
<div
  class="pt-4"
>
  <h6>
    Comments
  </h6>
  <div
    class="list-group"
  >
    <CommentListItem
      comment={
        Object {
          "commentBody": "this is my comment",
          "commenter": "Kylu",
          "createdAt": 125,
          "id": "1",
        }
      }
      key="1"
    />
    <CommentListItem
      comment={
        Object {
          "commentBody": "this is my another comment",
          "commenter": "Chris",
          "createdAt": 130,
          "id": "2",
        }
      }
      key="2"
    />
  </div>
</div>
`;
